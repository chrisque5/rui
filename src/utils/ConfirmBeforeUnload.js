import { useEffect, useCallback } from 'react';

const ConfirmBeforeUnload = () => {
  const onBeforeUnload = useCallback((e) => {
    e.preventDefault();
    e.returnValue = '';
  }, []);

  useEffect(() => {
    window.addEventListener('beforeunload', onBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
    };
  }, [onBeforeUnload]);

  return null;
};

export default ConfirmBeforeUnload;
