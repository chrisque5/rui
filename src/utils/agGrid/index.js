import {
  approvalsCellRenderer, approvalsSummaryCellRenderer, isUnderInvestigationCellRenderer, stpCellRenderer, stpFilterCellRenderer,
} from './js-renderers';
import {
  dateFormatter, formatDate, numberFormatter, timeFormatter,
  valueDateFormatter, fixingDateFormatter, publishDateFormatter,
  priceFormatter, brokerageFormatter,
} from './formatters';
import { dateFilterComparator, numberFilterComparitor, timeFilterComparator } from './comparitors';
import { getExportMenuItems } from './exportUtils';

export {
  approvalsCellRenderer,
  approvalsSummaryCellRenderer,
  dateFilterComparator,
  dateFormatter,
  formatDate,
  getExportMenuItems,
  isUnderInvestigationCellRenderer,
  numberFilterComparitor,
  numberFormatter,
  stpCellRenderer,
  stpFilterCellRenderer,
  timeFilterComparator,
  timeFormatter,
  valueDateFormatter,
  fixingDateFormatter,
  publishDateFormatter,
  priceFormatter,
  brokerageFormatter,
};
