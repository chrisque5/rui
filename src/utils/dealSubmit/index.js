import mapFormToFwdChain from './mapFormToFwdChain';
import mapFormToNdfChain from './mapFormToNdfChain';
import mapFormToSptChain from './mapFormToSptChain';

export { mapFormToFwdChain, mapFormToNdfChain, mapFormToSptChain };
