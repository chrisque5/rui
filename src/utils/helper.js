import moment from 'moment';
import * as math from 'mathjs';
import { Modal } from 'antd';

export const isValidTerm = (termToTest, selections) => {
  if (termToTest.length < 2) return false;

  const lowerCaseTermToTest = termToTest.toLowerCase();

  const endsWithPeriod = (term) => {
    const periods = ['d', 'w', 'm', 'y'];
    return (periods.includes(term.charAt(term.length - 1)));
  };

  const firstCharacterIsNumber = (term) => {
    const result = term.substring(0, term.length - 1).match(/^[0-9]*$/);
    return result ? result.length > 0 : false;
  };

  const isStaticSelectionValid = (term) => {
    const validStaticTerms = selections.map((x) => x.toLowerCase());
    if (validStaticTerms.includes(term)) {
      return true;
    }

    if (term.startsWith('bmf') || term.startsWith('imm')) {
      let termDaysOnly = term.replace('bmf', '');
      termDaysOnly = termDaysOnly.replace('imm', '');
      const beforeConversion = termDaysOnly;
      termDaysOnly = parseInt(termDaysOnly, 10);
      if (beforeConversion.length === termDaysOnly.toString().length) {
        return (termDaysOnly % 1 === 0 && termDaysOnly > 0 && termDaysOnly < 101);
      }
    }
    return false;
  };

  return isStaticSelectionValid(lowerCaseTermToTest)
    || (endsWithPeriod(lowerCaseTermToTest) && firstCharacterIsNumber(lowerCaseTermToTest));
};

export const doesArrayExistWithValue = (array, index = 0) => {
  if (Array.isArray(array)) {
    if (array && array.length > 0 && array[index]) {
      return true;
    }
  }
  return false;
};

export const setTodayOnNull = (form, id, value) => {
  if (!value) {
    setTimeout(() => {
      form.setFieldsValue({ [id]: moment() });
    });
  }
};

export const dayDifferenceIgnoreHours = (earlierDateMoment, laterDateMoment) => {
  const date1 = earlierDateMoment.clone().startOf('day');
  const date2 = laterDateMoment.clone().startOf('day');
  const dayCount = date1.diff(date2, 'days');

  return dayCount;
};

export const getDateInLocalTZ = () => moment().tz(moment.tz.guess(true)).format('YYYY-MM-DD');
export const getDateTimeInLocalTZ = () => moment().tz(moment.tz.guess(true));

export const parseValueAsBigNumber = (value) => {
  try {
    return math.bignumber(value);
  } catch (err) {
    return null;
  }
};

export const formatAmount = (val) => {
  const inputValue = val.replace(/\s/g, '');
  let multiplier = 1;
  const oneM = 1000000;
  const oneB = 1000000000;
  let amt = inputValue;

  if (Number.isNaN(Number(inputValue)) && inputValue.length > 1) {
    const lastChar = inputValue.charAt(inputValue.length - 1).toLowerCase();
    amt = Number(inputValue.substring(0, inputValue.length - 1));
    if (lastChar === 'm') {
      multiplier = oneM;
    } else if (lastChar === 'b') {
      multiplier = oneB;
    } else {
      return inputValue;
    }
  } else if (Number(inputValue) < oneM) {
    multiplier = oneM;
  }

  const calculatedAmount = math.multiply(math.bignumber(amt), math.bignumber(multiplier));
  return calculatedAmount.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
};

export const isValidDate = (date, dateFormat) => moment(date, dateFormat || 'L', true).isValid();

export const findCurrencyPair = (ccy1, ccy2, currencies) => (
  currencies.find(
    (currency) => currency.baseCurrency === ccy1 && currency.counterCurrency === ccy2,
  )
);

export const filterSpotExecutionVenues = (execVenues) => execVenues.filter((i) => i.venueId === 'XOFF');

export const isDealTypeSame = (dealType1, dealType2) => (dealType1 === dealType2);

export const isStrategyTypeSame = (strategyType1, strategyType2) => (strategyType1 === strategyType2);

export const ConfirmUnsavedChanges = (message, onYes, onNo) => {
  Modal.confirm({
    title: 'Confirm',
    content: message,
    okText: 'Yes',
    cancelText: 'No',
    onOk: () => {
      if (onYes) {
        setTimeout(() => {
          onYes();
        });
      }
    },
    onCancel: () => {
      if (onNo) {
        setTimeout(() => {
          onNo();
        });
      }
    },
  });
};

export const isObjectEmpty = (obj) => Object.keys(obj).length === 0 && obj.constructor === Object;

export const getUniqueID = () => {
  const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
  return `${s4() + s4()}-${s4()}`;
};

export const showPreviousValueToolTip = (currentValue, previousValue) => {
  if (currentValue !== previousValue) {
    return { title: `Previous Value: ${previousValue || 'N/A'}`, className: 'deal-input-dirty' };
  }
  return { title: '', className: '' };
};
