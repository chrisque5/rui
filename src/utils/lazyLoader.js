import { lazy } from 'react';

export const lazyWithPreload = (factory) => {
  const Component = lazy(factory);
  Component.preload = factory;
  return Component;
};

export const dummy = 'dummy';
