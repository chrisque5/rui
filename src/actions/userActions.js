import { actionTypes } from '../utils/constants';
import userApi from '../api/userApi';

export function loadUserSuccess(user) {
  return { type: actionTypes.LOAD_USER_SUCCESS, user };
}

export function loadUserFailed(user) {
  return { type: actionTypes.LOAD_USER_FAILED, user };
}

export function loadUser() {
  return ((dispatch) => userApi.getUser()
    .then((response) => dispatch(loadUserSuccess(response)))
    .catch((error) => dispatch(loadUserFailed(error)))
  );
}

export function loadPreferencesSuccess(preferences) {
  return { type: actionTypes.LOAD_PREFERENCES_SUCCESS, payload: { ...preferences } };
}

export function loadPreferencesFailed(error) {
  return { type: actionTypes.LOAD_PREFERENCES_FAILED, error };
}

export function loadPreferences() {
  return ((dispatch) => userApi.getPreferences()
    .then((response) => dispatch(loadPreferencesSuccess({ preferences: { ...response }, ...response })))
    .catch((error) => dispatch(loadPreferencesFailed(error)))
  );
}

export function preferencesChangeSuccess(preferences) {
  return { type: actionTypes.CHANGE_PREFERENCE_SUCCESS, payload: { preferences } };
}

export function preferencesChangeFailed(error) {
  return { type: actionTypes.CHANGE_PREFERENCE_FAILED, error };
}

export function deleteBrokerPreferenceSuccess(preferences) {
  return { type: actionTypes.DELETE_BROKER_PREFERENCE_SUCCESS, payload: { ...preferences } };
}

export function deleteBrokerPreferenceFailed(error) {
  return { type: actionTypes.DELETE_BROKER_PREFERENCE_FAILED, error };
}

export function addClientTraderPreferenceSuccess(payload) {
  return { type: actionTypes.ADD_CLIENTTRADER_PREFERENCE_SUCCESS, payload };
}

export function addClientTraderPreferenceFailed(error) {
  return { type: actionTypes.ADD_CLIENTTRADER_PREFERENCE_FAILED, error };
}

export function updateDefaultPreferencesAction(newDeal) {
  return { type: actionTypes.UPDATE_DEFAULTS, newDeal };
}

export function addClientTraderPreference(brokerDeskIds, clientTraderIds, agentId) {
  return async (dispatch, getState) => {
    try {
      const response = await userApi.addClientTraderPreference(brokerDeskIds, clientTraderIds, agentId);

      const state = getState();
      const { ui = {} } = state;
      const { selectedPreferenceFirm = null } = ui;
      const { preferredBrokers = [] } = response;
      const [deskId = null, id = null] = brokerDeskIds;

      const payload = {
        preferences: response, brokerDeskIds, preferredBrokers, brokerId: id, deskId, tradingCustomerId: selectedPreferenceFirm,
      };

      return dispatch(addClientTraderPreferenceSuccess(payload));
    } catch (error) {
      return dispatch(addClientTraderPreferenceFailed(error));
    }
  };
}

export function deleteClientTraderPreference(brokerDeskIds, clientData) {
  return ((dispatch) => userApi.deleteClientTraderPreference(brokerDeskIds, clientData)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function editClientTraderPreference(brokerDeskIds, clientTraderId) {
  return ((dispatch) => userApi.editClientTraderPreference(brokerDeskIds, clientTraderId)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function deleteBrokerPreference(brokerDeskIds) {
  return ((dispatch) => userApi.deleteBrokerPreference(brokerDeskIds)
    .then((response) => dispatch(deleteBrokerPreferenceSuccess({ preferences: { ...response }, ...response })))
    .catch((error) => dispatch(deleteBrokerPreferenceFailed(error)))
  );
}

export function editBrokerOrder(newOrder, existingOrder, favType, selectedBroker) {
  return ((dispatch) => userApi.editBrokerOrderPreference(
    newOrder, existingOrder, favType, selectedBroker,
  ).then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function addCurrencyPairPreference(brokerDeskIds, currency1, currency2, dealtCurrency) {
  return ((dispatch) => userApi.addCurrencyPairPreference(brokerDeskIds, currency1, currency2, dealtCurrency)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function deleteCurrencyPairPreference(brokerDeskIds, currencyPair) {
  return ((dispatch) => userApi.deleteCurrencyPairPreference(brokerDeskIds, currencyPair)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function addTermPreference(brokerDeskIds, term, valueDate) {
  return ((dispatch) => userApi.addTermPreference(brokerDeskIds, term, valueDate)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function deleteTermPreference(brokerDeskIds, term) {
  return ((dispatch) => userApi.deleteTermPreference(brokerDeskIds, term)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function changePreference(newSettings) {
  return ((dispatch) => userApi.modifyUserSettings(newSettings)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function addExecutionVenuePreference(brokerDeskIds, executionVenue) {
  return ((dispatch) => userApi.addExecutionVenuePreference(brokerDeskIds, executionVenue)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function deleteExecutionVenuePreference(brokerDeskIds, executionVenue) {
  return ((dispatch) => userApi.deleteExecutionVenuePreference(brokerDeskIds, executionVenue)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function updateDefaultPreferences(newDeal) {
  return ((dispatch) => userApi.updateDefaults(newDeal)
    .then((response) => dispatch(preferencesChangeSuccess(response)))
    .catch((error) => dispatch(preferencesChangeFailed(error)))
  );
}

export function initUserData() {
  return (dispatch) => dispatch(loadPreferences()).then(() => dispatch(loadUser()));
}

/**
 * When a firm has been selected, load the clients for that firm
 */
export function selectPreferredFirmComplete(payload) {
  return { type: actionTypes.SELECT_PREFERRED_FIRM_COMPLETE, payload };
}

export function selectPreferredFirm(selectedFirm = {}, brokerDeskIds) {
  return (dispatch, getState) => {
    const state = getState();
    const { user: { preferences: { preferredBrokers } = {} } = {} } = state;
    const [deskId, brokerId] = brokerDeskIds;
    const { tradingCustomerCodeLocationCode } = selectedFirm;

    const payload = {
      preferredBrokers, brokerId, deskId, tradingCustomerCodeLocationCode,
    };

    dispatch(selectPreferredFirmComplete(payload));
  };
}
