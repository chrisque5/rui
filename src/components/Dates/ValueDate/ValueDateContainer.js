import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Modal } from 'antd';
import moment from 'moment';
import ValueDate from './ValueDate';
import * as userActions from '../../../actions/userActions';
import { ids } from '../../../utils/constants';
import { getIsDateResponsePending } from '../../../utils/selectors';

class ValueDateContainer extends Component {
  addToPreference = () => {
    const valueDate = moment(this.props.form.getFieldValue(`${ids.VALUE_DATE_1}`)).format('L');
    const selectedBrokerDeskId = this.props.userInterface.selectedPreferenceBroker;
    if (!selectedBrokerDeskId || selectedBrokerDeskId === '') {
      Modal.warning({
        title: 'Warning',
        content: 'No Broker tab open to add the favourite. Please open a Broker tab.',
      });
    } else if (valueDate) {
      this.props.userActions.addTermPreference(selectedBrokerDeskId, null, valueDate);
    }
  };

  valueDateOnChange = (value, valueDateId) => {
    // no local onChange events here.
    this.props.onChangeAfterEffects(value, valueDateId);
  }

  render() {
    return (
      <ValueDate
        form={this.props.form}
        id={this.props.id}
        label={this.props.label}
        testId={this.props.testId}
        valueDateOnChange={this.valueDateOnChange}
        addPreferenceClick={this.addToPreference}
        spotDate={this.props.spotDate}
        isPending={this.props.isDateResponsePending}
      />
    );
  }
}

ValueDateContainer.defaultProps = {
  label: '',
  onChangeAfterEffects: () => { },
  spotDate: null,
};

ValueDateContainer.propTypes = {
  form: PropTypes.shape().isRequired,
  spotDate: PropTypes.shape(),
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  testId: PropTypes.string.isRequired,
  onChangeAfterEffects: PropTypes.func,
  userActions: PropTypes.shape().isRequired,
  userInterface: PropTypes.shape().isRequired,
  isDateResponsePending: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
  userInterface: state.ui,
  isDateResponsePending: getIsDateResponsePending(state),
});

const mapDispatchToProps = (dispatch) => ({
  userActions: bindActionCreators(userActions, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(ValueDateContainer);
