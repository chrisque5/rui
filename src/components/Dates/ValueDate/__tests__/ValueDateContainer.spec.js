import moment from 'moment';
import { Modal } from 'antd';
import {
  renderWithForm, waitFor, getByTitle, fireEvent,
} from 'test-utils/testUtils';
import ValueDateContainer from '../ValueDateContainer';
import { ids } from '../../../../utils/constants';
import userActions from '../../../../actions/userActions';

jest.mock('../../../../utils/helper', () => ({ setTodayOnNull: jest.fn() }));
jest.mock('moment', () => () => jest.requireActual('moment')('2020-04-04T00:00:00.000Z'));
jest.mock('../../../../actions/userActions', () => ({
  addTermPreference: jest.fn(() => jest.fn(() => Promise.resolve())),
}));

describe('<ValueDateContainer/>', () => {
  const mockForm = {
    getFieldValue: (field) => {
      if (field === ids.VALUE_DATE_1) {
        return moment();
      }
      return 'dummy';
    },
    setFieldValue: (field) => field,

    resetFields: () => { },
  };

  const modalSpy = jest.spyOn(Modal, 'warning');
  const onChangeAfterEffects = jest.fn();
  const addTermPreferenceSpy = jest.spyOn(userActions, 'addTermPreference');

  const defaultProps = {
    form: mockForm,
    id: ids.VALUE_DATE_1,
    onChangeAfterEffects,
    initialValue: moment(),
    testId: 'dtpValueDate1',
    label: 'Value Date',
  };
  let defaultState = {
    ui: {
      selectedPreferenceBroker: '',
    },
  };

  const renderComponent = (initialState) => renderWithForm(<ValueDateContainer {...defaultProps} />, { initialState });

  test('renders snapshot as expected', () => {
    const { container } = renderComponent(defaultState);
    expect(container).toMatchSnapshot();
  });

  test('verify valueDateOnChange', async () => {
    const { getByTestId } = renderComponent(defaultState);

    const dateInput = await waitFor(() => getByTestId('dtpValueDate1'));
    fireEvent.mouseDown(dateInput);

    const selectedDateElem = await waitFor(() => getByTitle(document.body, '2020-04-11'));
    fireEvent.click(selectedDateElem);

    expect(onChangeAfterEffects).toHaveBeenCalledTimes(1);
  });

  test('verify onclick event on star without broker', async () => {
    const { getByTestId } = renderComponent(defaultState);
    const star = await waitFor(() => getByTestId('addPreference'));
    fireEvent.click(star);
    expect(modalSpy).toHaveBeenCalledTimes(1);
  });

  test('verify onclick event on star with broker', async () => {
    defaultState = {
      ui: {
        selectedPreferenceBroker: {},
      },
    };
    const { getByTestId } = renderComponent(defaultState);
    const star = await waitFor(() => getByTestId('addPreference'));
    fireEvent.click(star);
    expect(addTermPreferenceSpy).toHaveBeenCalledTimes(1);
  });
});
