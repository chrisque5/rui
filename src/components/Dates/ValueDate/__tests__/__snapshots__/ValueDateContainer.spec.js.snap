// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ValueDateContainer/> renders snapshot as expected 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item item-value-date"
    >
      <div
        class="ant-col ant-form-item-label"
      >
        <label
          class="ant-form-item-required"
          for="valueDate1"
          title=""
        >
          Value Date
          <span
            aria-label="star"
            class="anticon anticon-star star-icon"
            data-testid="addPreference"
            role="img"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="star"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M908.1 353.1l-253.9-36.9L540.7 86.1c-3.1-6.3-8.2-11.4-14.5-14.5-15.8-7.8-35-1.3-42.9 14.5L369.8 316.2l-253.9 36.9c-7 1-13.4 4.3-18.3 9.3a32.05 32.05 0 00.6 45.3l183.7 179.1-43.4 252.9a31.95 31.95 0 0046.4 33.7L512 754l227.1 119.4c6.2 3.3 13.4 4.4 20.3 3.2 17.4-3 29.1-19.5 26.1-36.9l-43.4-252.9 183.7-179.1c5-4.9 8.3-11.3 9.3-18.3 2.7-17.5-9.5-33.7-27-36.3z"
              />
            </svg>
          </span>
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-picker"
              style="width: 100%;"
            >
              <div
                class="ant-picker-input"
              >
                <input
                  autocomplete="off"
                  data-testid="dtpValueDate1"
                  id="valueDate1"
                  placeholder="Select date"
                  readonly=""
                  size="12"
                  tabindex="-1"
                  title="04/04/2020"
                  value="04/04/2020"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
